central/notifier/processor/singleton.go

```diff
@@ -111,34 +108,41 @@ func initialize() {
 				err = notifierUtils.RekeyNotifier(protoNotifier, oldKey, cryptoKey)
 				if err != nil {
 					utils.Should(fmt.Errorf("error rekeying notifier %s, notifications to this notifier will fail", protoNotifier.GetId()))
-					continue
+					return
 				}
 				notifiersToUpsert = append(notifiersToUpsert, protoNotifier)
 			}
 		}
-
-		err = notifierDatastore.UpsertManyNotifiers(ctx, notifiersToUpsert)
-		if err != nil {
-			utils.Should(errors.Wrap(err, "Error upserting secured notifiers, several notifiers will be unable to send notifications"))
-		}
 		if needsRekey {
-			// If we did a rekey, then update the stored key index
-			encConfig = &storage.NotifierEncConfig{ActiveKeyIndex: int32(activeIndex)}
-			err = encConfigDataStore.UpsertConfig(encConfig)
-			if err != nil {
-				utils.Should(errors.Wrapf(err, "Error updating notifier encryption config's stored key index to %d", activeIndex))
-			}
+			defer func() {
+				// If we did a rekey, then update the stored key index
+				encConfig := &storage.NotifierEncConfig{ActiveKeyIndex: int32(activeIndex)}
+				err = encConfigDataStore.UpsertConfig(encConfig)
+				if err != nil {
+					utils.Should(errors.Wrapf(err, "Error updating notifier encryption config's stored key index to %d", activeIndex))
+				}
+
+			}()
 		}
 	}
 
+	if err := notifierDatastore.UpsertManyNotifiers(ctx, notifiersToUpsert); err != nil {
+		utils.Should(errors.Wrap(err, "Error upserting secured notifiers, several notifiers will be unable to send notifications"))
+	}
```

Comment:
IMHO, this should happen after the foreach loop. Otherwise, there will be no notifier to upsert.